onnx/defs/schema.h

```diff
@@ -205,6 +207,9 @@ class OpSchema {
   OpSchema& EnforceConsumed(std::unordered_map<int, int> inplace);
   OpSchema& EnforceOneToOneConsumed();
 
+  OpSchema& TensorInferenceFunction(std::function<void(Node*)> inferenceFunction);
```

Comment:
Dumb questions: The Node does not contain input tensor data, right? if yes, input tensor data may be needed in some cases to calculate the outputs' shape.

Reply:
yes, there can be data dependent ops. But the point of shape inference is to be able to operate on GraphProto only - i.e. with shapes but without actual data. Thus it might produce "undefined" shapes along the way.

@anderspapitto - let's document it somewhere in comments

Reply:
@linkerzhang in general, inputs are not available, but those that are initializers to the graph are (see addInitializer in ir.h)

@dzhulgakov "undefined" shapes isn't anything special, it's basically the default. But sure, I can add a comment that shape inference is best-effort and not guaranteed to figure out everything